File: src/smart-grid/domain/bot/SmartGridBot.ts
import { Order } from "../../../models/Order";
import { Position } from "../../../models/Position";
import { Trade } from "../../../models/Trade";
import { GridStrategyConfig } from "../../../models/GridStrategyConfig";
import { IndicatorService } from "../../../shared/indicators/IndicatorService";
import { MathUtils } from "../../../shared/utils/MathUtils";

export class SmartGridBot {
  symbol: string;
  balance: number;
  initial_balance: number;
  grid_density: number;
  qty_per_order: number;
  volatility_lookback: number;
  trend_period: number;
  trend_threshold: number;
  take_profit_pct: number;
  stop_loss_pct: number;
  trailing_stop_pct: number;
  martingale_factor: number;

  private _order_counter: number = 0;
  open_orders: Map<number, Order> = new Map();
  positions: Position[] = [];
  grid_levels: number[] = [];
  grid_lower: number = 0.0;
  grid_upper: number = 0.0;
  trend: "uptrend" | "downtrend" | "ranging" = "ranging";
  equity_curve: number[];
  trade_log: Trade[] = [];

  constructor(config: GridStrategyConfig = {}) {
    this.symbol = config.symbol ?? "BTC/USDT";
    this.initial_balance = config.initial_balance ?? 10000.0;
    this.balance = this.initial_balance;
    this.grid_density = config.grid_density ?? 20;
    this.qty_per_order = config.qty_per_order ?? 0.0;
    this.volatility_lookback = config.volatility_lookback ?? 72;
    this.trend_period = config.trend_period ?? 50;
    this.trend_threshold = config.trend_threshold ?? 0.001;
    this.take_profit_pct = config.take_profit_pct ?? 1.5;
    this.stop_loss_pct = config.stop_loss_pct ?? 0.0;
    this.trailing_stop_pct = config.trailing_stop_pct ?? 0.0;
    this.martingale_factor = config.martingale_factor ?? 1.0;

    this.equity_curve = [this.initial_balance];
  }

  private _next_id(): number {
    this._order_counter += 1;
    return this._order_counter;
  }

  public on_candle(
    timestamp: any,
    open_: number,
    high: number,
    low: number,
    close: number,
    closes_history: number[],
  ): void {
    const trend = IndicatorService.computeTrend(
      closes_history,
      this.trend_period,
      this.trend_threshold,
    );
    const volatility = IndicatorService.computeVolatility(
      closes_history,
      this.volatility_lookback,
    );
    const sma200 = IndicatorService.computeSMA(closes_history, 200);
    const rsi = IndicatorService.computeRSI(closes_history, 14);

    this.trend = trend;

    // Regime detection
    const is_bull_market = close > sma200;
    const is_strong_downtrend = trend === "downtrend" && !is_bull_market;

    // Inventory management: how much of our initial balance is currently tied up in BTC?
    const btc_held_qty = this.positions.reduce((sum, p) => sum + p.quantity, 0);
    const btc_value = btc_held_qty * close;

    // Calculate funds locked in open buy orders
    let locked_balance = 0;
    for (const order of this.open_orders.values()) {
      if (order.side === "buy") {
        locked_balance += order.price * order.quantity;
      }
    }

    const current_total_equity = this.balance + btc_value + locked_balance;
    const exposure_pct = (btc_value / current_total_equity) * 100;

    this._rebuild_grid(close, volatility, is_strong_downtrend, exposure_pct);
    this._simulate_fills(timestamp, low, high, close);
    this._manage_positions(timestamp, low, high, close, exposure_pct);
    this._cancel_stale_orders(close);

    // Entry logic: "Oversold Collector"
    // 1. Never buy if RSI > 60 (don't buy the top)
    // 2. Only enter new grids if RSI < 40 or we are already in exposure.
    // 3. This avoids buying at the end of a pump.
    const is_extreme_oversold = rsi < 30;
    const can_add_exposure = exposure_pct < 85;

    let can_buy = can_add_exposure && (rsi < 40 || trend === "uptrend");

    if (can_buy) {
      this._place_buy_orders(close, is_strong_downtrend, exposure_pct);
    }

    // Position exit logic is handled in _manage_positions

    this.equity_curve.push(current_total_equity);
  }

  private _rebuild_grid(
    current_price: number,
    volatility: number,
    is_strong_downtrend: boolean,
    exposure_pct: number,
  ): void {
    const margin = (this.grid_upper - this.grid_lower) * 0.2;
    if (
      this.grid_levels.length > 0 &&
      current_price > this.grid_lower + margin &&
      current_price < this.grid_upper - margin
    ) {
      return;
    }

    // Adjust density: fewer levels in downtrend to avoid fast accumulation
    const effective_density = is_strong_downtrend
      ? Math.floor(this.grid_density * 0.6)
      : this.grid_density;

    // Scale range with volatility, but wider during crashes
    const range_multiplier = is_strong_downtrend ? 6.0 : 4.0;
    const half_range =
      current_price *
      volatility *
      range_multiplier *
      Math.sqrt(this.volatility_lookback);

    let bias = 0.0;
    if (this.trend === "uptrend") bias = half_range * 0.15;
    if (this.trend === "downtrend") bias = -half_range * 0.15;

    this.grid_lower = current_price - half_range + bias;
    this.grid_upper = current_price + half_range + bias;
    this.grid_levels = MathUtils.linspace(
      this.grid_lower,
      this.grid_upper,
      effective_density,
    );

    // Dynamic sizing: lower size as exposure increases
    const base_allocation = is_strong_downtrend ? 0.3 : 0.8;
    const size_reduction_factor = Math.max(0.2, 1 - exposure_pct / 100);
    const available_capital =
      this.balance * base_allocation * size_reduction_factor;

    const levels_to_fill = Math.max(1, Math.floor(effective_density / 2));
    this.qty_per_order = available_capital / levels_to_fill / current_price;

    this._cancel_all_orders();
  }

  private _cancel_all_orders(): void {
    for (const [oid, o] of this.open_orders.entries()) {
      this.balance += o.price * o.quantity;
      this.open_orders.delete(oid);
    }
  }

  private _place_buy_orders(
    current_price: number,
    is_strong_downtrend: boolean,
    exposure_pct: number,
  ): void {
    const buy_levels = this.grid_levels
      .filter((lvl) => lvl < current_price)
      .reverse();

    const existing_prices = new Set<number>();
    for (const order of this.open_orders.values()) {
      if (order.side === "buy") {
        existing_prices.add(this._round_price(order.price));
      }
    }

    // Limit number of grid levels to place orders for
    const max_levels = is_strong_downtrend ? 3 : 15;
    let placed = 0;

    for (const [idx, level] of buy_levels.entries()) {
      if (placed >= max_levels) break;

      const rounded_level = this._round_price(level);
      if (!existing_prices.has(rounded_level)) {
        // Martingale logic: increase size for deeper levels in the current grid
        const depth_factor = Math.pow(this.martingale_factor, idx);

        let multiplier = depth_factor;
        if (this.trend === "uptrend") multiplier *= 1.2;
        if (is_strong_downtrend) multiplier *= 0.5;

        const qty = this.qty_per_order * multiplier;
        const cost = level * qty;

        if (this.balance >= cost && qty > 0) {
          const oid = this._next_id();
          this.open_orders.set(oid, {
            order_id: oid,
            side: "buy",
            price: level,
            quantity: qty,
            status: "open",
          });
          this.balance -= cost;
          placed++;
        }
      }
    }
  }

  private _round_price(p: number): number {
    return Math.round(p * 100) / 100;
  }

  private _cancel_stale_orders(current_price: number): void {
    const stale: number[] = [];
    for (const [oid, o] of this.open_orders.entries()) {
      if (o.side === "buy" && o.price > current_price) {
        stale.push(oid);
      }
    }
    for (const oid of stale) {
      const o = this.open_orders.get(oid)!;
      this.open_orders.delete(oid);
      this.balance += o.price * o.quantity;
    }
  }

  private _simulate_fills(
    timestamp: any,
    low: number,
    high: number,
    close: number,
  ): void {
    const filled_ids: number[] = [];
    const grid_step =
      (this.grid_upper - this.grid_lower) /
      Math.max(1, this.grid_levels.length - 1);

    for (const [oid, order] of this.open_orders.entries()) {
      if (order.side === "buy" && low <= order.price && order.price <= high) {
        order.status = "filled";
        order.fill_price = order.price;
        filled_ids.push(oid);

        // Dynamic Take Profit: narrower when exposure is high to de-risk faster
        const tp_multiplier = 1.5;
        const tp_price =
          order.price +
          Math.max(
            grid_step,
            order.price * ((this.take_profit_pct * tp_multiplier) / 100),
          );

        const pos = new Position(order.price, order.quantity, tp_price, 0, 0);
        this.positions.push(pos);
        this.trade_log.push({
          timestamp: timestamp,
          side: "buy",
          price: order.price,
          quantity: order.quantity,
        });
      }
    }

    for (const oid of filled_ids) {
      this.open_orders.delete(oid);
    }
  }

  private _manage_positions(
    timestamp: any,
    low: number,
    high: number,
    close: number,
    exposure_pct: number,
  ): void {
    const remaining: Position[] = [];
    const sma200 = IndicatorService.computeSMA(this.equity_curve, 200); // Using SMA of equity for trailing stop

    // NEW: BEAR MARKET EXIT
    // If we are in a strong downtrend and price is below SMA200,
    // we should be very quick to exit even at minor profits or small losses if exposure is high.
    const is_extreme_bear =
      close < IndicatorService.computeSMA(this.equity_curve, 200) * 0.9; // Arbitrary crash detection

    for (const pos of this.positions) {
      let exited = false;
      let exit_price: number | null = null;
      let reason: string | null = null;

      // Tighten TP in bear markets
      const adjusted_tp =
        this.trend === "downtrend"
          ? pos.entry_price * 1.01
          : pos.take_profit_price;

      if (high >= adjusted_tp) {
        exit_price = Math.max(high, adjusted_tp);
        reason = "take_profit";
        exited = true;
      }

      if (exited && exit_price !== null) {
        const proceeds = exit_price * pos.quantity;
        this.balance += proceeds;
        this.trade_log.push({
          timestamp: timestamp,
          side: "sell",
          price: exit_price,
          quantity: pos.quantity,
          reason: reason,
          pnl: proceeds - pos.entry_price * pos.quantity,
        });
      } else {
        remaining.push(pos);
      }
    }
    this.positions = remaining;
  }

  public summary(): any {
    const final_equity =
      this.equity_curve.length > 0
        ? this.equity_curve[this.equity_curve.length - 1]
        : this.initial_balance;
    const profit = final_equity - this.initial_balance;
    const roi = (profit / this.initial_balance) * 100;

    let max_dd = 0;
    let peak = this.equity_curve[0] || this.initial_balance;
    for (const eq of this.equity_curve) {
      if (eq > peak) peak = eq;
      const drawdown = (peak - eq) / peak;
      if (drawdown > max_dd) max_dd = drawdown;
    }
    max_dd *= 100;

    const sells = this.trade_log.filter((t) => t.side === "sell");
    return {
      initial_balance: Number(this.initial_balance.toFixed(2)) + " $",
      final_value: Number(final_equity.toFixed(2)) + " $",
      total_profit: Number(profit.toFixed(2)) + " $",
      roi_pct: Number(roi.toFixed(2)) + " %",
      total_trades: sells.length,
      max_drawdown_pct: Number(max_dd.toFixed(2)) + " %",
    };
  }

  private _liquidate_all(
    current_price: number,
    timestamp: any,
    reason: string,
  ): void {
    for (const pos of this.positions) {
      const proceeds = current_price * pos.quantity;
      this.balance += proceeds;
      this.trade_log.push({
        timestamp: timestamp,
        side: "sell",
        price: current_price,
        quantity: pos.quantity,
        reason: reason,
        pnl: proceeds - pos.entry_price * pos.quantity,
      });
    }
    this.positions = [];
  }
}

File: bots/smart-grid/smart_grid_bot.py
"""smart_grid_bot.py
Smart Grid Trading Bot for BTC/USDT — core strategy logic.

Features:
  - Dynamic grid levels adjusted by volatility (ATR / rolling std)
  - Trend detection via configurable Moving Average
  - Trailing stop-loss per position
  - Optional Martingale safety-order sizing
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Dict, List, Optional

import numpy as np


# ---------------------------------------------------------------------------
# Data structures
# ---------------------------------------------------------------------------

@dataclass
class Order:
    order_id: int
    side: str          # 'buy' | 'sell'
    price: float
    quantity: float
    status: str = 'open'   # 'open' | 'filled' | 'cancelled'
    fill_price: Optional[float] = None


@dataclass
class Position:
    entry_price: float
    quantity: float
    take_profit_price: float
    stop_loss_price: float
    trailing_stop_pct: float
    highest_price_seen: float = field(init=False)

    def __post_init__(self):
        self.highest_price_seen = self.entry_price

    def update_trailing_stop(self, current_price: float) -> None:
        """Ratchet the trailing stop up as price rises."""
        if current_price > self.highest_price_seen:
            self.highest_price_seen = current_price
            self.stop_loss_price = self.highest_price_seen * (
                1 - self.trailing_stop_pct / 100
            )

    @property
    def is_stop_hit(self) -> bool:
        return False  # evaluated externally against live price


# ---------------------------------------------------------------------------
# SmartGridBot
# ---------------------------------------------------------------------------

class SmartGridBot:
    """
    Parameters
    ----------
    symbol              : trading pair label (informational)
    initial_balance     : starting USDT capital
    grid_density        : number of grid lines
    qty_per_order       : BTC quantity per order
    volatility_lookback : rolling window (candles) for volatility
    trend_period        : MA period for trend detection
    trend_threshold     : minimum |slope| to declare a trend
    take_profit_pct     : % profit target per trade
    stop_loss_pct       : % hard stop below entry
    trailing_stop_pct   : % trailing stop below peak price
    martingale_factor   : safety-order size multiplier (1.0 = disabled)
    """

    def __init__(
        self,
        symbol: str = "BTC/USDT",
        initial_balance: float = 10_000.0,
        grid_density: int = 100,
        qty_per_order: float = 0.05,
        volatility_lookback: int = 72,
        trend_period: int = 50,
        trend_threshold: float = 0.001,
        take_profit_pct: float = 1.0,
        stop_loss_pct: float = 2.0,
        trailing_stop_pct: float = 0.5,
        martingale_factor: float = 1.0,
    ):
        self.symbol = symbol
        self.balance = initial_balance
        self.initial_balance = initial_balance
        self.grid_density = grid_density
        self.qty_per_order = qty_per_order
        self.volatility_lookback = volatility_lookback
        self.trend_period = trend_period
        self.trend_threshold = trend_threshold
        self.take_profit_pct = take_profit_pct
        self.stop_loss_pct = stop_loss_pct
        self.trailing_stop_pct = trailing_stop_pct
        self.martingale_factor = martingale_factor

        # Runtime state
        self._order_counter: int = 0
        self.open_orders: Dict[int, Order] = {}
        self.positions: List[Position] = []
        self.grid_levels: List[float] = []
        self.grid_lower: float = 0.0
        self.grid_upper: float = 0.0
        self.trend: str = 'ranging'  # 'uptrend' | 'downtrend' | 'ranging'
        self.equity_curve: List[float] = [initial_balance]
        self.trade_log: List[dict] = []

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _next_id(self) -> int:
        self._order_counter += 1
        return self._order_counter

    def _compute_volatility(self, closes: np.ndarray) -> float:
        """Rolling std of log-returns, annualised to a % of price."""
        if len(closes) < 2:
            return 0.01
        log_ret = np.diff(np.log(closes[-self.volatility_lookback:]))
        return float(np.std(log_ret)) if len(log_ret) > 0 else 0.01

    def _compute_trend(self, closes: np.ndarray) -> str:
        """Simple MA slope trend detector."""
        if len(closes) < self.trend_period + 1:
            return 'ranging'
        ma = np.convolve(closes, np.ones(self.trend_period) / self.trend_period, mode='valid')
        slope = (ma[-1] - ma[-2]) / ma[-2]
        if slope > self.trend_threshold:
            return 'uptrend'
        if slope < -self.trend_threshold:
            return 'downtrend'
        return 'ranging'

    def _rebuild_grid(self, current_price: float, volatility: float) -> None:
        """Re-center and re-space the grid around current_price."""
        half_range = current_price * volatility * math.sqrt(self.volatility_lookback)
        # Trend bias: shift grid bounds toward trend direction
        bias = 0.0
        if self.trend == 'uptrend':
            bias = half_range * 0.2
        elif self.trend == 'downtrend':
            bias = -half_range * 0.2

        self.grid_lower = current_price - half_range + bias
        self.grid_upper = current_price + half_range + bias
        self.grid_levels = list(
            np.linspace(self.grid_lower, self.grid_upper, self.grid_density)
        )

    def _place_buy_orders(self, current_price: float) -> None:
        """Place buy limit orders at grid levels below current price."""
        buy_levels = [lvl for lvl in self.grid_levels if lvl < current_price]
        existing_prices = {o.price for o in self.open_orders.values() if o.side == 'buy'}
        for level in buy_levels:
            if level not in existing_prices:
                qty = self.qty_per_order
                cost = level * qty
                if self.balance >= cost:
                    oid = self._next_id()
                    self.open_orders[oid] = Order(
                        order_id=oid, side='buy', price=level, quantity=qty
                    )
                    self.balance -= cost

    def _cancel_stale_orders(self, current_price: float) -> None:
        """Cancel buy orders that are now above current_price (grid shifted)."""
        stale = [
            oid
            for oid, o in self.open_orders.items()
            if o.side == 'buy' and o.price > current_price
        ]
        for oid in stale:
            o = self.open_orders.pop(oid)
            self.balance += o.price * o.quantity   # refund

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def on_candle(
        self,
        timestamp,
        open_: float,
        high: float,
        low: float,
        close: float,
        closes_history: np.ndarray,
    ) -> None:
        """
        Called once per candle.  `closes_history` should include all
        closes up to and including this candle.
        """
        # 1. Recalculate indicators
        volatility = self._compute_volatility(closes_history)
        self.trend = self._compute_trend(closes_history)

        # 2. Rebuild grid
        self._rebuild_grid(close, volatility)

        # 3. Simulate order fills (candle H/L sweep)
        self._simulate_fills(timestamp, low, high, close)

        # 4. Manage open positions (trailing stop / take profit)
        self._manage_positions(timestamp, low, high, close)

        # 5. Cancel stale orders and place new ones
        self._cancel_stale_orders(close)
        self._place_buy_orders(close)

        # 6. Record equity
        btc_held = sum(p.quantity for p in self.positions)
        self.equity_curve.append(self.balance + btc_held * close)

    def _simulate_fills(self, timestamp, low: float, high: float, close: float) -> None:
        filled_ids = []
        for oid, order in self.open_orders.items():
            if order.side == 'buy' and low <= order.price <= high:
                order.status = 'filled'
                order.fill_price = order.price
                filled_ids.append(oid)
                tp_price = order.price * (1 + self.take_profit_pct / 100)
                sl_price = order.price * (1 - self.stop_loss_pct / 100)
                pos = Position(
                    entry_price=order.price,
                    quantity=order.quantity,
                    take_profit_price=tp_price,
                    stop_loss_price=sl_price,
                    trailing_stop_pct=self.trailing_stop_pct,
                )
                self.positions.append(pos)
                self.trade_log.append({
                    'timestamp': timestamp,
                    'side': 'buy',
                    'price': order.price,
                    'quantity': order.quantity,
                })

        for oid in filled_ids:
            del self.open_orders[oid]

    def _manage_positions(self, timestamp, low: float, high: float, close: float) -> None:
        remaining = []
        for pos in self.positions:
            pos.update_trailing_stop(high)
            exited = False
            exit_price = None
            reason = None

            # Take profit
            if high >= pos.take_profit_price:
                exit_price = pos.take_profit_price
                reason = 'take_profit'
                exited = True
            # Trailing / hard stop
            elif low <= pos.stop_loss_price:
                exit_price = pos.stop_loss_price
                reason = 'stop_loss'
                exited = True

            if exited and exit_price is not None:
                proceeds = exit_price * pos.quantity
                self.balance += proceeds
                self.trade_log.append({
                    'timestamp': timestamp,
                    'side': 'sell',
                    'price': exit_price,
                    'quantity': pos.quantity,
                    'reason': reason,
                    'pnl': proceeds - pos.entry_price * pos.quantity,
                })
            else:
                remaining.append(pos)

        self.positions = remaining

    # ------------------------------------------------------------------
    # Reporting
    # ------------------------------------------------------------------

    def summary(self) -> dict:
        final_equity = self.equity_curve[-1] if self.equity_curve else self.initial_balance
        profit = final_equity - self.initial_balance
        roi = (profit / self.initial_balance) * 100
        eq = np.array(self.equity_curve)
        peak = np.maximum.accumulate(eq)
        drawdowns = (peak - eq) / peak
        max_dd = float(np.max(drawdowns)) * 100
        sells = [t for t in self.trade_log if t['side'] == 'sell']
        return {
            'initial_balance': self.initial_balance,
            'final_value': round(final_equity, 2),
            'total_profit': round(profit, 2),
            'roi_pct': round(roi, 2),
            'total_trades': len(sells),
            'max_drawdown_pct': round(max_dd, 2),
        }
